### PR TITLE
fix: keep message while backoff stash buffer unstash exception (#31814)

### PR DESCRIPTION
**Pull request by Roiocam**
References #31814

Keep message while unStash exception on backoff SupervisorStrategy.

### CODE REVIEW

**Comment by Roiocam** [comment]
close-reopen to retrigger PR unrelated CI failed.

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/AskSpec.scala [hunk 1]
```diff
@@ -119,17 +121,19 @@ class AskSpec extends ScalaTestWithActorTestKit("""
...
 
+      val actor: ActorRef[Msg] = spawn(behavior)
       val deadLetterProbe = createDeadLetterProbe()
-      val probe = createTestProbe[Any]()
-      actor ! Stop(probe.ref)
-      implicit val timeout: Timeout = 10.millis
+      val terminated: Future[Unit] = actor.ask(Stop(_))
+      Await.result(terminated, 100.millis)
+      // give actor some time to terminated.
+      Thread.sleep(100)
```
Comment (Roiocam) [inline akka-actor-typed-tests/src/test/scala/akka/actor/typed/AskSpec.scala:131]: fix not robustness on #32104

Comment (leviramsey) [reply akka-actor-typed-tests/src/test/scala/akka/actor/typed/AskSpec.scala:131]: Using the probe to block the test until termination has propagated would be more efficient than a 100ms sleep... in fact this test case can be merged with the case starting at line 60 which tests the same scenario (ask of an already-terminated actor).

Comment (Roiocam) [reply akka-actor-typed-tests/src/test/scala/akka/actor/typed/AskSpec.scala]: Great input of test, thanks!

Comment (Roiocam) [reply akka-actor-typed-tests/src/test/scala/akka/actor/typed/AskSpec.scala]: Solved on #32111

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/SupervisionSpec.scala [hunk 2]
```diff
@@ -310,6 +310,27 @@ class SupervisionSpec extends ScalaTestWithActorTestKit("""
       }).onFailure[IllegalArgumentException](strategy)
   }
 
+  class FailingRestartBackoffTestSetup(strategy: SupervisorStrategy) {
+    val probe = TestProbe[AnyRef]("evt")
+
+    def behv =
+      supervise(setup[Command] { context =>
+        // inject some delay on here.
+        Thread.sleep(200)
```
Comment (patriknw) [inline akka-actor-typed-tests/src/test/scala/akka/actor/typed/SupervisionSpec.scala:319]: Why is this needed? Is it to fill up the mailbox with all `msg` so that they are added to the stash when throw?
Can that be replaced with a CountDownLatch instead to make the test more deterministic?

**Review by patriknw — COMMENTED** [review]
looking good

**Review by patriknw — APPROVED** [review]
LGTM

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala [hunk 3]
```diff
@@ -354,9 +354,15 @@ private class RestartSupervisor[T, Thr <: Throwable: ClassTag](initial: Behavior
...
-      (StashBuffer[Any](ctx.asScala.asInstanceOf[scaladsl.ActorContext[Any]], stashCapacity), childrenToStop))
-
+    // new generate only if there has been reset to None.
+    restartingInProgress = restartingInProgress match {
+      case OptionVal.Some((stashBuffer, _)) =>
+        // keep stashBuffer on here.
+        OptionVal.Some((stashBuffer, childrenToStop))
+      case _ =>
+        OptionVal.Some(
+          (StashBuffer[Any](ctx.asScala.asInstanceOf[scaladsl.ActorContext[Any]], stashCapacity), childrenToStop))
```
Comment (johanandren) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala:364]: I think we could simplify this to only cover the part that differs, something like:

```suggestion
    // new stash only if there is no already on-going restart with previously stashed messages
    val stashBufferForRestart = restartingInProgress match {
      case OptionVal.Some((stashBuffer, _)) => stashBuffer
      case _ => StashBuffer[Any](ctx.asScala.asInstanceOf[scaladsl.ActorContext[Any]], stashCapacity)
    restartingInProgress = OptionVal.Some((stashBufferForRestart, childrenToStop))
```

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala [hunk 4]
```diff
@@ -387,8 +393,10 @@ private class RestartSupervisor[T, Thr <: Throwable: ClassTag](initial: Behavior
       val newBehavior = Behavior.validateAsInitial(Behavior.start(initial, ctx.asInstanceOf[TypedActorContext[T]]))
       val nextBehavior = restartingInProgress match {
         case OptionVal.Some((stashBuffer, _)) =>
+          val behavior = stashBuffer.unstashAll(newBehavior.unsafeCast)
+          // reset to None only if this completed on success
```
Comment (johanandren) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala:397]:
```suggestion
          // restart unstash was successful for all stashed messages, drop stash buffer
```

**Review by johanandren — COMMENTED** [review]
Two tiny improvement suggestions and then I think this is good to go

**Review by johanandren — APPROVED** [review]
LGTM, thanks!
